### PR TITLE
Adding possibility to use single line drawing when displaying the dependency tree

### PR DESCRIPTION
**Pull request by kurazsi**
## The original tree drawing example:
```
+-<itemObject>
  +-<itemObject>
  | +-<itemObject>
  | | +-<itemObject>
  | | | +-<itemObject>
  | | |   +-<itemObject>
  | | |   | +-<itemObject>
  | | |   | +-<itemObject>
  | | |   | +-<itemObject>
  | | |   | +-<itemObject>
  | | |   | +-<itemObject>
  | | |   | +-<itemObject>

...
```
## The new proposed option:
By adding to the command line argument the extra flag "linedraw" like this:  --tree=linedraw,
the print will be generated in this format:
```
└─┬<itemObject>
  ├─┬<itemObject>
  │ ├─┬<itemObject>
  │ │ └─┬<itemObject>
  │ │   ├─┬<itemObject>
  │ │   │ └─┬<itemObject>
  │ │   │   ├─┬<itemObject>
  │ │   │   │ ├─<itemObject>
  │ │   │   │ ├─<itemObject>
  │ │   │   │ ├─<itemObject>
  │ │   │   │ ├─<itemObject>
  │ │   │   │ ├─<itemObject>
  │ │   │   │ ├─<itemObject>
  │ │   │   │ ├─<itemObject>
  │ │   │   │ ├─<itemObject>
  │ │   │   │ ├─<itemObject>
  │ │   │   │ ├─<itemObject>
  │ │   │   │ ├─<itemObject>
  │ │   │   │ └─<itemObject>
  │ │   │   └─┬<itemObject>
  │ │   │     ├─<itemObject>

...
```

## Contributor Checklist:

* [x] I have created a new test or updated the unit tests to cover the new/changed functionality.
* [x] I have updated `src/CHANGES.txt`  (and read the `README.txt` in that directory)
* [ ] I have updated the appropriate documentation

### CODE REVIEW

**Comment by tipeipaul** [comment]
This is the original drawing:

+-validation/validationApps/testApp_1
  +-validation/validationApps/testApp_1
    +-validation/validationApps/testApp_1/app.elf
      +-validation/validationApps/testApp_1/auxiliaryFunctions.lib
      | +-validation/validationApps/testApp_1/common.lib
      |   +-validation/validationApps/system/libfile00.a
      |   | +-validation/validationApps/system/objectfile00.c.o
      |   | +-validation/validationApps/system/objectfile01.c.o
      |   | +-validation/validationApps/system/objectfile02.c.o
      |   | +-validation/validationApps/system/objectfile03.c.o
      |   | +-validation/validationApps/system/objectfile04.c.o
      |   | +-validation/validationApps/system/objectfile05.c.o
      |   | +-validation/validationApps/system/objectfile06.c.o
      |   | +-validation/validationApps/system/objectfile07.c.o
      |   | +-validation/validationApps/system/objectfile08.c.o
      |   | +-validation/validationApps/system/objectfile09.c.o
      |   | +-validation/validationApps/system/objectfile10.c.o
      |   | +-validation/validationApps/system/objectfile11.c.o
      |   +-validation/validationApps/system/libfile01.a
      |     +-validation/validationApps/system/objectfile12.c.o
      |     +-validation/validationApps/system/objectfile12.c.o
      |     +-validation/validationApps/system/objectfile12.c.o
      |     +-validation/validationApps/system/objectfile12.c.o
      |     +-validation/validationApps/system/objectfile12.c.o
      |     +-validation/validationApps/system/objectfile12.c.o
      |     +-validation/validationApps/system/objectfile12.c.o
      |     +-validation/validationApps/system/objectfile12.c.o
      |     +-validation/validationApps/system/objectfile12.c.o
      |     +-validation/validationApps/system/objectfile12.c.o
      |     +-validation/validationApps/system/objectfile12.c.o
      |     +-validation/validationApps/system/objectfile12.c.o
      |     +-validation/validationApps/system/objectfile12.c.o
      |     +-validation/validationApps/system/objectfile12.c.o
      |     +-validation/validationApps/system/objectfile12.c.o
      +-validation/validationApps/testApp_1/main_app.lib
        +-validation/validationApps/testApp_1/main_op1.a
        | +-validation/validationApps/testApp_1/main_opAdd.a
        |   +-validation/validationApps/testApp_1/main_opAdd.c.o
        +-validation/validationApps/testApp_1/main_op2.a
        | +-validation/validationApps/testApp_1/main_opSub.a
        |   +-validation/validationApps/testApp_1/main_opSub.c.o
        +-[validation/validationApps/testApp_1/ext_op1.a]
        +-[validation/validationApps/testApp_1/ext_op2.a]

**Comment by tipeipaul** [comment]
+-validation/validationApps/testApp_1
  +-validation/validationApps/testApp_1
    +-validation/validationApps/testApp_1/app.elf
      +-validation/validationApps/testApp_1/auxiliaryFunctions.lib
      | +-validation/validationApps/testApp_1/common.lib
      |   +-validation/validationApps/system/libfile00.a
      |   | +-validation/validationApps/system/objectfile00.c.o
      |   | +-validation/validationApps/system/objectfile01.c.o
      |   | +-validation/validationApps/system/objectfile02.c.o
      |   | +-validation/validationApps/system/objectfile03.c.o
      |   | +-validation/validationApps/system/objectfile04.c.o
      |   | +-validation/validationApps/system/objectfile05.c.o
      |   | +-validation/validationApps/system/objectfile06.c.o
      |   | +-validation/validationApps/system/objectfile07.c.o
      |   | +-validation/validationApps/system/objectfile08.c.o
      |   | +-validation/validationApps/system/objectfile09.c.o
      |   | +-validation/validationApps/system/objectfile10.c.o
      |   | +-validation/validationApps/system/objectfile11.c.o
      |   +-validation/validationApps/system/libfile01.a
      |     +-validation/validationApps/system/objectfile12.c.o
      |     +-validation/validationApps/system/objectfile12.c.o
      |     +-validation/validationApps/system/objectfile12.c.o
      |     +-validation/validationApps/system/objectfile12.c.o
      |     +-validation/validationApps/system/objectfile12.c.o
      |     +-validation/validationApps/system/objectfile12.c.o
      |     +-validation/validationApps/system/objectfile12.c.o
      |     +-validation/validationApps/system/objectfile12.c.o
      |     +-validation/validationApps/system/objectfile12.c.o
      |     +-validation/validationApps/system/objectfile12.c.o
      |     +-validation/validationApps/system/objectfile12.c.o
      |     +-validation/validationApps/system/objectfile12.c.o
      |     +-validation/validationApps/system/objectfile12.c.o
      |     +-validation/validationApps/system/objectfile12.c.o
      |     +-validation/validationApps/system/objectfile12.c.o
      +-validation/validationApps/testApp_1/main_app.lib
        +-validation/validationApps/testApp_1/main_op1.a
        | +-validation/validationApps/testApp_1/main_opAdd.a
        |   +-validation/validationApps/testApp_1/main_opAdd.c.o
        +-validation/validationApps/testApp_1/main_op2.a
        | +-validation/validationApps/testApp_1/main_opSub.a
        |   +-validation/validationApps/testApp_1/main_opSub.c.o
        +-[validation/validationApps/testApp_1/ext_op1.a]
        +-[validation/validationApps/testApp_1/ext_op2.a]

**Comment by tipeipaul** [comment]
+-validation/validationApps/testApp_1
  +-validation/validationApps/testApp_1
    +-validation/validationApps/testApp_1/app.elf
      +-validation/validationApps/testApp_1/auxiliaryFunctions.lib
      | +-validation/validationApps/testApp_1/common.lib
      |   +-validation/validationApps/system/libfile00.a
      |   | +-validation/validationApps/system/objectfile00.c.o
      |   | +-validation/validationApps/system/objectfile01.c.o
      |   | +-validation/validationApps/system/objectfile02.c.o
      |   | +-validation/validationApps/system/objectfile03.c.o
      |   | +-validation/validationApps/system/objectfile04.c.o
      |   | +-validation/validationApps/system/objectfile05.c.o
      |   | +-validation/validationApps/system/objectfile06.c.o
      |   | +-validation/validationApps/system/objectfile07.c.o
      |   | +-validation/validationApps/system/objectfile08.c.o
      |   | +-validation/validationApps/system/objectfile09.c.o
      |   | +-validation/validationApps/system/objectfile10.c.o
      |   | +-validation/validationApps/system/objectfile11.c.o
      |   +-validation/validationApps/system/libfile01.a
      |     +-validation/validationApps/system/objectfile12.c.o
      |     +-validation/validationApps/system/objectfile12.c.o
      |     +-validation/validationApps/system/objectfile12.c.o
      |     +-validation/validationApps/system/objectfile12.c.o
      |     +-validation/validationApps/system/objectfile12.c.o
      |     +-validation/validationApps/system/objectfile12.c.o
      |     +-validation/validationApps/system/objectfile12.c.o
      |     +-validation/validationApps/system/objectfile12.c.o
      |     +-validation/validationApps/system/objectfile12.c.o
      |     +-validation/validationApps/system/objectfile12.c.o
      |     +-validation/validationApps/system/objectfile12.c.o
      |     +-validation/validationApps/system/objectfile12.c.o
      |     +-validation/validationApps/system/objectfile12.c.o
      |     +-validation/validationApps/system/objectfile12.c.o
      |     +-validation/validationApps/system/objectfile12.c.o
      +-validation/validationApps/testApp_1/main_app.lib
        +-validation/validationApps/testApp_1/main_op1.a
        | +-validation/validationApps/testApp_1/main_opAdd.a
        |   +-validation/validationApps/testApp_1/main_opAdd.c.o
        +-validation/validationApps/testApp_1/main_op2.a
        | +-validation/validationApps/testApp_1/main_opSub.a
        |   +-validation/validationApps/testApp_1/main_opSub.c.o
        +-[validation/validationApps/testApp_1/ext_op1.a]
        +-[validation/validationApps/testApp_1/ext_op2.a]

**Comment by bdbaddog** [comment]
not sure that linedraw should be an alternative to prune or all, you may want line draw with either prune or all.

**Comment by tipeipaul** [comment]
+-validation/validationApps/testApp_1
      +-validation/validationApps/testApp_1
        +-validation/validationApps/testApp_1/app.elf
          +-validation/validationApps/testApp_1/auxiliaryFunctions.lib
          | +-validation/validationApps/testApp_1/common.lib
          |   +-validation/validationApps/system/libfile00.a
          |   | +-validation/validationApps/system/objectfile00.c.o
          |   | +-validation/validationApps/system/objectfile01.c.o
          |   | +-validation/validationApps/system/objectfile02.c.o
          |   | +-validation/validationApps/system/objectfile03.c.o
          |   | +-validation/validationApps/system/objectfile04.c.o
          |   | +-validation/validationApps/system/objectfile05.c.o
          |   | +-validation/validationApps/system/objectfile06.c.o
          |   | +-validation/validationApps/system/objectfile07.c.o
          |   | +-validation/validationApps/system/objectfile08.c.o
          |   | +-validation/validationApps/system/objectfile09.c.o
          |   | +-validation/validationApps/system/objectfile10.c.o
          |   | +-validation/validationApps/system/objectfile11.c.o
          |   +-validation/validationApps/system/libfile01.a
          |     +-validation/validationApps/system/objectfile12.c.o
          |     +-validation/validationApps/system/objectfile12.c.o
          |     +-validation/validationApps/system/objectfile12.c.o
          |     +-validation/validationApps/system/objectfile12.c.o
          |     +-validation/validationApps/system/objectfile12.c.o
          |     +-validation/validationApps/system/objectfile12.c.o
          |     +-validation/validationApps/system/objectfile12.c.o
          |     +-validation/validationApps/system/objectfile12.c.o
          |     +-validation/validationApps/system/objectfile12.c.o
          |     +-validation/validationApps/system/objectfile12.c.o
          |     +-validation/validationApps/system/objectfile12.c.o
          |     +-validation/validationApps/system/objectfile12.c.o
          |     +-validation/validationApps/system/objectfile12.c.o
          |     +-validation/validationApps/system/objectfile12.c.o
          |     +-validation/validationApps/system/objectfile12.c.o
          +-validation/validationApps/testApp_1/main_app.lib
           +-validation/validationApps/testApp_1/main_op1.a
           | +-validation/validationApps/testApp_1/main_opAdd.a
           |   +-validation/validationApps/testApp_1/main_opAdd.c.o
          +-validation/validationApps/testApp_1/main_op2.a
           | +-validation/validationApps/testApp_1/main_opSub.a
           |   +-validation/validationApps/testApp_1/main_opSub.c.o
           +-[validation/validationApps/testApp_1/ext_op1.a]
           +-[validation/validationApps/testApp_1/ext_op2.a]

**Comment by bdbaddog** [comment]
Please update the blurb at the top with a description, and actually that's the better place to put the sample outputs.

**Comment by kurazsi** [comment]
> not sure that linedraw should be an alternative to prune or all, you may want line draw with either prune or all.

it's not an alternative, it just allow enabling that mode or using the old. Because we cant test it now on other platforms then posix

**Comment by bdbaddog** [comment]
> > not sure that linedraw should be an alternative to prune or all, you may want line draw with either prune or all.
> 
> it's not an alternative, it just allow enabling that mode or using the old. Because we cant test it now on other platforms then posix

Ah.. My bad. I misread the code. so you can do ```--tree=prune,linedraw``` ?
So you don't have access to any windows systems?

**Comment by kurazsi** [comment]
> > > not sure that linedraw should be an alternative to prune or all, you may want line draw with either prune or all.
> > 
> > 
> > it's not an alternative, it just allow enabling that mode or using the old. Because we cant test it now on other platforms then posix
> 
> Ah.. My bad. I misread the code. so you can do `--tree=prune,linedraw` ?
> So you don't have access to any windows systems?

Unfortunately not. Soon...

File: src/CHANGES.txt [hunk 1]
```diff
@@ -10,18 +10,22 @@ NOTE: Please include a reference to any Issues resolved by your changes in the b
 
 RELEASE  VERSION/DATE TO BE FILLED IN LATER
 
+  From Iosif Kurazs:
```
Comment (bdbaddog) [inline src/CHANGES.txt]: You must've missed this bit at the top of this file:
**NOTE: Please add your name below in alphabetical order by last name.**

**Comment by bdbaddog** [comment]
Please read the link above in the PR template: 

> ## Remove this paragraph
> Please have a look at our developer documentation before submitting your Pull Request.
> 
> https://scons.org/guidelines.html

Then delete it from the description text.

You'll need to add at least one test for this functionality.
Note when you add the tests and push to your branch the CI will run and test on linux and windows.

**Comment by tipeipaul** [comment]
> Please read the link above in the PR template:
> 
> > ## Remove this paragraph
> > Please have a look at our developer documentation before submitting your Pull Request.
> > https://scons.org/guidelines.html
> 
> Then delete it from the description text.
> 
> You'll need to add at least one test for this functionality.
> Note when you add the tests and push to your branch the CI will run and test on linux and windows.

Is there any test that tests the printing of the dependency tree?

**Comment by mwichmann** [comment]
> Is there any test that tests the printing of the dependency tree?

At least `test/option--tree.py` and the contents of `test/option/tree`.  There may be some other tests that have expectations on the output, I'll update if I find some.

File: test/option--tree.py [hunk 2]
```diff
@@ -76,6 +76,39 @@
...
+    # Python 2
+    write = unichr(0xe7).encode('utf-8')
+except NameError:
+    # Python 3
+    # str is utf-8 by default
+    write = chr(0xe7)
+env.Textfile("LineDraw", write)
+""")
+
+if sys.version_info.major < 3:
```
Comment (mwichmann) [inline test/option--tree.py]: you don't need this, scons master has no Python2 support any longer,

File: src/CHANGES.txt (hunk 1 above)
Comment (tipeipaul) [reply src/CHANGES.txt]: Thank you. I changed.

**Comment by mwichmann** [comment]
This change affects only `SCons.Util.print_tree`, not the related `SCons.Util.render_tree`. Is that the intended outcome?

**Comment by bdbaddog** [comment]
@mwichmann completed the doc changes and a few changes in implementation which has been merged in #3591 ..
Closing this PR.
Thanks for the great work!

**Pull request closed without merging**